File: prss.py
class PrSS:
    """Single-Row BMS"""
    def __init__(self, array):
        if not isinstance(array, list):
            raise TypeError(f'Expected list type for PrSS(), not {type(array)}')

        sequence = array[:]
        for i in range(len(sequence)):
            if not isinstance(sequence[i], int):
                raise TypeError(f'Expected int type for array[{i}], not {type(sequence[i])}')

        if len(sequence) == 0:
            self.sequence = sequence
            return

        last = -1
        for this in sequence:
            if this > (last + 1):
                raise ValueError(f'Invalid PrSS: {sequence}')
            last = this

        self.sequence = sequence

    def __str__(self):
        return ''.join(f'({item})' for item in self.sequence)

    def eval(self, base=3):
        """Evaluate a PrSS in a single level of recursion."""

        if not self.sequence:
            raise ValueError("Cannot evaluate an empty sequence.")

        if self.sequence[-1] == 0:
            s = self.sequence[:-1]
            return PrSS(s)

        bad = self.sequence[-1]
        s = []
        for i in range(len(self.sequence) - 2, -1, -1):
            s.insert(0, self.sequence[i])
            if self.sequence[i] < bad:
                t = self.sequence[:i]
                return PrSS(t + s * (base + 1))
        return PrSS(s * base)

File: js/prss.js
function repeatArray(arr, count) { 
    let result = []; 
    for (let i = 0; i < count; i++) { 
        result = result.concat(arr); 
    } 
    return result; 
}

class PrSS {
    constructor(array) {
        if (!Array.isArray(array)) {
            throw new TypeError(`Expected list type for PrSS(), not ${typeof array}`);
        }

        this.sequence = array.slice();
        for (let i = 0; i < this.sequence.length; i++) {
            if (typeof this.sequence[i] !== 'number') {
                throw new TypeError(`Expected int type for array[${i}], not ${typeof this.sequence[i]}`);
            }
        }

        if (this.sequence.length === 0) {
            return;
        }

        let last = -1;
        for (const value of this.sequence) {
            if (value > (last + 1)) {
                throw new Error(`Invalid PrSS: ${this.sequence}`);
            }
            last = value;
        }
    }

    toString() {
        return this.sequence.map(item => `(${item})`).join('');
    }

    eval(base = 3) {
        if (this.sequence.length === 0) {
            throw new Error("Cannot evaluate an empty sequence.");
        }

        if (this.sequence[this.sequence.length - 1] === 0) {
            const s = this.sequence.slice(0, -1);
            return new PrSS(s);
        }

        const bad = this.sequence[this.sequence.length - 1];
        let s = [];
        for (let i = this.sequence.length - 2; i >= 0; i--) {
            s.unshift(this.sequence[i]);
            if (this.sequence[i] < bad) {
                const t = this.sequence.slice(0, i);
                return new PrSS(t.concat(repeatArray(s, base + 1));
            }
        }
        return new PrSS(repeatArray(s, base);
    }
}

module.exports = PrSS;
